### PR TITLE
an es modules only re-factor. fixes #516   DO NOT MERGE THIS

### LINKED ISSUE #516: Snabbdom and ECMAScript Modules
I think it's time to discuss: Snabbdom and ECMAScript Modules.

My thoughts: Snabbdom should be standard-compliant as much as possible.
UPD: About current state of the art, see for instance [1](https://2ality.com/2019/04/nodejs-esm-impl.html), [2](https://2ality.com/2019/10/hybrid-npm-packages.html).

All issues about this ATM: #200, #225, #247, #261, #312, #319, #338, #437, #522, #523, #542, #543, cyclejs/cyclejs#925, #568, #601.

/cc @paldepind , @mightyiam , @nojvek

### PR DESCRIPTION
**Pull request by mreinstein**
changes:
* removes commonjs build
* removes browserify/bundling from all examples, instead loading them as `script type="module"`
* modified `tsconfig.json` to generate es6 (it was set to es5)
* updated the README to only use esm syntax
* adds a `"type": "modules",` field to `package.json`
* adds an `"exports": { ... }` field to `package.json`
* updated several dot files to reference node v14.1.0 rather than node 12

things I didn't do:
* I made no attempts at re-organizing the way things are exported (I know there's discussion/polite argument about named vs default exports and I haven't touched any of that)
* I'm not sure the karma build is working correctly. I'm guessing something is just broken with the config, due to my karma related ignorance. :)
* didn't remove webpack usage (I think we could, but wasn't sure if karma is relying on it)

### CODE REVIEW

**Comment by kuraga** [comment]
> I made no attempts at re-organizing the way things are exported (I know there's discussion/polite argument about named vs default exports and I haven't touched any of that)

After solving, might this change?
```js
import * as snabbdom from '../../es/snabbdom.js';
```

File: README.md [hunk 1]
```diff
@@ -58,30 +58,40 @@ performance, small size and all the features listed below.
...
+import * as snabbdom from 'snabbdom';
+import clazz         from 'snabbdom/class';
+import h             from 'snabbdom/h';
+import props         from 'snabbdom/props';
+import style         from 'snabbdom/style';
+import listeners     from 'snabbdom/eventlisteners';
+
+
+// Init patch function with chosen modules
+var patch = init([
```
Comment (kuraga) [inline README.md:70]:
```suggestion
var patch = snabbdom.init([
```

Comment (kuraga) [inline README.md:70]:
```suggestion
var patch = snabbdom.init([
```

**Comment by kuraga** [comment]
@mreinstein , :+1: :+1: :+1:

**Comment by mreinstein** [comment]
> I would split this into several pull requests.

I could remove the node version bump, but I'm unclear what effect that would have on the build? The other changes in the PR rely on being at node 14 and up.


> I would leave `.js` files as `.js` for now.

Can you elaborate on this? I left the source files as `.js`, but I changed the  eslint and karma files from .`js` -> `.cjs`, in the hopes that they would continue to work. I'm not sure if you're referring to the source files or the config files in your comment about file extensions.

**Comment by mreinstein** [comment]
Is there another branch I should send this PR to besides master? I forked the repo and created a new branch, `esm-next`.

File: package.json [hunk 2]
```diff
@@ -2,48 +2,58 @@
...
+    "karma-mocha-reporter": "^2.2.5",
+    "karma-webpack": "^4.0.2",
+    "lodash.shuffle": "^4.2.0",
+    "mocha": "^7.1.2",
+    "npm-run-all": "^4.1.5",
+    "remark-cli": "^7.0.1",
+    "remark-toc": "^7.0.0",
+    "tsconfigs": "^4.0.2",
+    "typescript": "^3.8.3",
+    "webpack": "^4.43.0"
```
Comment (mightyiam) [inline package.json:56]: 1. The versions are pinned on purpose. We use Renovate to keep them up-to-date. In theory. In practice, #595 is making this tedious.
1. If you bump any packages, please explain why that is necessary, and please bump only those packages.

File: package.json [hunk 3]
```diff
@@ -53,9 +63,7 @@
     "lint": "run-s lint:editorconfig lint:js",
     "unit": "karma start",
     "test": "run-s lint compile unit",
-    "compile": "npm run compile-es && npm run compile-commonjs",
-    "compile-es": "tsc --outDir es --module es6",
-    "compile-commonjs": "tsc --outDir . --module commonjs",
+    "compile": "tsc --outDir es --module es6",
```
Comment (mightyiam) [inline package.json:66]: If `--outDir` and `--module` are static, then they could be in the `tsconfig.json`.

File: src/h.ts [hunk 4]
```diff
@@ -1,5 +1,5 @@
-import { vnode, VNode, VNodeData } from './vnode';
-import * as is from './is';
+import { vnode, VNode, VNodeData } from './vnode.js';
+import * as is from './is.js';
```
Comment (mightyiam) [inline src/h.ts:2]: Is this recommended practice, please?

I know that extensionless import paths are non-ES-modules-spec-compliant. This doesn't seem right, though.

**Comment by mightyiam** [comment]
> @mightyiam
> 
> > I would split this into several pull requests.
> 
> I could remove the node version bump, but I'm unclear what effect that would have on the build?

Not for the purpose of having any affect on the build. In fact, I didn't even look at the build log. For the purpose of separating changes.

> The other changes in the PR rely on being at node 14 and up.

What changes are those that rely on Node.js v14 and up, please? The source code runs solely in the browser during tests, doesn't it?

> 
> > I would leave `.js` files as `.js` for now.
> 
> Can you elaborate on this? I left the source files as `.js`

But our source files are `.ts`, aren't they?

> , but I changed the eslint and karma files from .`js` -> `.cjs`, in the hopes that they would continue to work.

With the belief that this change is required for them to continue to work in Node.js v14, or for some other purpose, and with the hope that they would continue to work?

> I'm not sure if you're referring to the source files or the config files in your comment about file extensions.

I am referring to the config files.

> Is there another branch I should send this PR to besides master? I forked the repo and created a new branch, `esm-next`.

`master` would be the correct target. Breaking changes are long overdue, I feel.

**Comment by mreinstein** [comment]
0k, thanks for clarifying. I will close this PR, and send a clean one which:
* removes the `.js` -> `.cjs` file renames
* removes the package.json dep updates
* removes the version bumps in nodenv, etc.

File: src/h.ts (hunk 4 above)
Comment (mreinstein) [reply src/h.ts:2]: Can you elaborate? I'm not sure what you mean by not right. Are you saying they should be including `.ts` versions instead?

Comment (kuraga) [reply src/h.ts:2]: If you're talking about `*` then https://github.com/snabbdom/snabbdom/pull/592#issuecomment-622307219, I think.

**Comment by mreinstein** [comment]
@mightyiam are you on slack or discord somewhere? Maybe it would be easier to chat in real time. I think that might be quicker

**Comment by mightyiam** [comment]
@mreinstein next time, instead of closing an existing PR and opening a new one, reuse the same PR. Rewrite its commit history.

> @mightyiam are you on slack or discord somewhere? Maybe it would be easier to chat in real time. I think that might be quicker.
> 
> I'll be here if you want to chat really quick: http://18.207.227.126:5000/

I feel that asynchronous communication is generally better practice for code reviews.

**Pull request closed without merging**